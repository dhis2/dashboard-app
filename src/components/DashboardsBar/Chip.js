import { useDhis2ConnectionStatus } from '@dhis2/app-runtime'
import { Chip as UiChip, colors, IconStarFilled24 } from '@dhis2/ui'
import cx from 'classnames'
import debounce from 'lodash/debounce.js'
import PropTypes from 'prop-types'
import React from 'react'
import { Link } from 'react-router-dom'
import { apiPostDataStatistics } from '../../api/dataStatistics.js'
import { useCacheableSection } from '../../modules/useCacheableSection.js'
import { OfflineSaved } from './assets/icons.js'
import classes from './styles/Chip.module.css'

const Chip = ({ starred, selected, label, dashboardId, onClick }) => {
    const { lastUpdated } = useCacheableSection(dashboardId)
    const { isConnected: online } = useDhis2ConnectionStatus()
    const chipProps = {
        selected,
    }

    if (starred) {
        chipProps.icon = (
            <IconStarFilled24
                color={selected ? colors.white : colors.grey600}
            />
        )
    }
    const debouncedPostStatistics = debounce(
        () => apiPostDataStatistics('DASHBOARD_VIEW', dashboardId),
        500
    )

    const handleClick = () => {
        online && debouncedPostStatistics()
        onClick()
    }

    return (
        <Link
            className={classes.link}
            to={`/${dashboardId}`}
            onClick={handleClick}
            data-test="dashboard-chip"
        >
            <UiChip {...chipProps}>
                <span
                    className={
                        lastUpdated ? classes.labelWithAdornment : undefined
                    }
                >
                    {label}
                </span>
                {lastUpdated && (
                    <OfflineSaved
                        className={cx(
                            classes.adornment,
                            selected && classes.selected
                        )}
                    />
                )}
            </UiChip>
        </Link>
    )
}

Chip.propTypes = {
    dashboardId: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    selected: PropTypes.bool.isRequired,
    starred: PropTypes.bool.isRequired,
    onClick: PropTypes.func.isRequired,
}

export default Chip
