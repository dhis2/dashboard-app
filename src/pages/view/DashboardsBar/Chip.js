import React from 'react'
import PropTypes from 'prop-types'
import { Chip as UiChip, colors, IconStarFilled24 } from '@dhis2/ui'
import { Link } from 'react-router-dom'
import debounce from 'lodash/debounce'

import { apiPostDataStatistics } from '../../../api/dataStatistics'

import classes from './styles/Chip.module.css'

export const Chip = ({ starred, selected, label, dashboardId, onClick }) => {
    const chipProps = {
        selected,
    }

    if (starred) {
        chipProps.icon = (
            <IconStarFilled24
                color={selected ? colors.white : colors.grey600}
            />
        )
    }
    const debouncedPostStatistics = debounce(
        () => apiPostDataStatistics('DASHBOARD_VIEW', dashboardId),
        500
    )

    const handleClick = () => {
        debouncedPostStatistics()
        onClick()
    }

    return (
        <Link
            className={classes.link}
            to={`/${dashboardId}`}
            onClick={handleClick}
            data-test="dashboard-chip"
        >
            <UiChip {...chipProps}>{label}</UiChip>
        </Link>
    )
}

Chip.propTypes = {
    dashboardId: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    selected: PropTypes.bool.isRequired,
    starred: PropTypes.bool.isRequired,
    onClick: PropTypes.func.isRequired,
    classes: PropTypes.object,
}

export default Chip
