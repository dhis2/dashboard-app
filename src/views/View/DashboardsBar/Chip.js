import React from 'react'
import PropTypes from 'prop-types'
import { Chip as UiChip } from '@dhis2/ui'
import { Link } from 'react-router-dom'
import debounce from 'lodash/debounce'

import StarIcon from './assets/Star'
import { apiPostDataStatistics } from '../../../api/dataStatistics'

import classes from './styles/Chip.module.css'

export const Chip = ({ starred, selected, label, dashboardId, onClick }) => {
    const chipProps = {
        selected,
    }

    if (starred) {
        const selectedState = selected ? classes.selected : classes.unselected
        chipProps.icon = (
            <StarIcon className={`${classes.icon} ${selectedState}`} />
        )
    }
    const debouncedPostStatistics = debounce(
        () => apiPostDataStatistics('DASHBOARD_VIEW', dashboardId),
        500
    )

    const handleClick = () => {
        debouncedPostStatistics()
        onClick()
    }

    return (
        <Link
            className={classes.link}
            to={`/${dashboardId}`}
            onClick={handleClick}
            data-test="dashboard-chip"
        >
            <UiChip {...chipProps}>{label}</UiChip>
        </Link>
    )
}

Chip.propTypes = {
    dashboardId: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    selected: PropTypes.bool.isRequired,
    starred: PropTypes.bool.isRequired,
    onClick: PropTypes.func.isRequired,
    classes: PropTypes.object,
}

export default Chip
