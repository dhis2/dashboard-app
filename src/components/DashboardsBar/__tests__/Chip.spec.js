import { useCacheableSection } from '@dhis2/app-runtime'
import { render } from '@testing-library/react'
import { createMemoryHistory } from 'history'
import React from 'react'
import { Router } from 'react-router-dom'
import Chip from '../Chip.js'

jest.mock('@dhis2/app-runtime', () => ({
    useOnlineStatus: () => ({ online: true }),
    useCacheableSection: jest.fn(),
}))

jest.mock('@dhis2/app-runtime-adapter-d2', () => ({
    useD2: () => ({
        d2: {
            currentUser: {
                username: 'rainbowDash',
                id: 'r3nb0d5h',
            },
        },
    }),
}))

const mockOfflineDashboard = {
    lastUpdated: 'Jan 10',
}

const mockNonOfflineDashboard = {
    lastUpdated: null,
}

const defaultProps = {
    starred: false,
    selected: false,
    onClick: jest.fn(),
    label: 'Rainbow Dash',
    dashboardId: 'rainbowdash',
    classes: {
        icon: 'iconClass',
        selected: 'selectedClass',
        unselected: 'unselectedClass',
    },
}

test('renders an unstarred chip for a non-offline dashboard', () => {
    useCacheableSection.mockImplementation(() => mockNonOfflineDashboard)
    const { container } = render(
        <Router history={createMemoryHistory()}>
            <Chip {...defaultProps} />
        </Router>
    )

    expect(container).toMatchSnapshot()
})

test('renders an unstarred chip for an offline dashboard', () => {
    useCacheableSection.mockImplementation(() => mockOfflineDashboard)
    const { container } = render(
        <Router history={createMemoryHistory()}>
            <Chip {...defaultProps} />
        </Router>
    )

    expect(container).toMatchSnapshot()
})

test('renders a starred chip for a non-cached dashboard', () => {
    useCacheableSection.mockImplementation(() => mockNonOfflineDashboard)
    const props = Object.assign({}, defaultProps, { starred: true })
    const { container } = render(
        <Router history={createMemoryHistory()}>
            <Chip {...props} />
        </Router>
    )

    expect(container).toMatchSnapshot()
})

test('renders a starred chip for a cached dashboard', () => {
    useCacheableSection.mockImplementation(() => mockOfflineDashboard)
    const props = Object.assign({}, defaultProps, { starred: true })
    const { container } = render(
        <Router history={createMemoryHistory()}>
            <Chip {...props} />
        </Router>
    )

    expect(container).toMatchSnapshot()
})

test('renders a starred, selected chip for non-cached dashboard', () => {
    useCacheableSection.mockImplementation(() => mockNonOfflineDashboard)
    const props = Object.assign({}, defaultProps, {
        starred: true,
        selected: true,
    })
    const { container } = render(
        <Router history={createMemoryHistory()}>
            <Chip {...props} />
        </Router>
    )

    expect(container).toMatchSnapshot()
})

test('renders a starred, selected chip for a cached dashboard', () => {
    useCacheableSection.mockImplementation(() => mockOfflineDashboard)
    const props = Object.assign({}, defaultProps, {
        starred: true,
        selected: true,
    })
    const { container } = render(
        <Router history={createMemoryHistory()}>
            <Chip {...props} />
        </Router>
    )

    expect(container).toMatchSnapshot()
})
