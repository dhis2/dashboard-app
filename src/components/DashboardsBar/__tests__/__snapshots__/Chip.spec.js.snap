// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a starred chip for a cached dashboard 1`] = `
<div>
  <a
    class="link"
    data-test="dashboard-chip"
    href="/rainbowdash"
  >
    <span
      class="jsx-3285323350 "
      data-test="dhis2-uicore-chip"
    >
      <span
        class="jsx-3489677657"
        data-test="dhis2-uicore-chip-icon"
      >
        <svg
          color="#6e7a8a"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.397 7.041l-6.051.88-.114.022a1 1 0 00-.44 1.683l4.378 4.268-1.033 6.027-.013.11a1 1 0 001.463.944L12 18.13l5.413 2.845.1.047a1 1 0 001.35-1.1l-1.034-6.028 4.38-4.268.078-.086a1 1 0 00-.633-1.62l-6.052-.879-2.705-5.484a1 1 0 00-1.794 0z"
            fill="currentColor"
          />
        </svg>
      </span>
      <span
        class="jsx-785997980 "
      >
        <span
          class="labelWithAdornment"
        >
          Rainbow Dash
        </span>
        <svg
          class="adornment"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m4.5 7v1h-1.5v6h10v-6h-1.5v-1h1.5c.5522847 0 1 .44771525 1 1v6c0 .5522847-.4477153 1-1 1h-10c-.55228475 0-1-.4477153-1-1v-6c0-.55228475.44771525-1 1-1zm5 3v1h-3v-1zm2.1130214-9.31661889.7739572.63323779-4.84985989 5.92760659-3.3906721-3.3906721.70710678-.70710678 2.60944661 2.60955339z"
            fill="inherit"
            fill-rule="evenodd"
          />
        </svg>
      </span>
    </span>
  </a>
</div>
`;

exports[`renders a starred chip for a non-cached dashboard 1`] = `
<div>
  <a
    class="link"
    data-test="dashboard-chip"
    href="/rainbowdash"
  >
    <span
      class="jsx-3285323350 "
      data-test="dhis2-uicore-chip"
    >
      <span
        class="jsx-3489677657"
        data-test="dhis2-uicore-chip-icon"
      >
        <svg
          color="#6e7a8a"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.397 7.041l-6.051.88-.114.022a1 1 0 00-.44 1.683l4.378 4.268-1.033 6.027-.013.11a1 1 0 001.463.944L12 18.13l5.413 2.845.1.047a1 1 0 001.35-1.1l-1.034-6.028 4.38-4.268.078-.086a1 1 0 00-.633-1.62l-6.052-.879-2.705-5.484a1 1 0 00-1.794 0z"
            fill="currentColor"
          />
        </svg>
      </span>
      <span
        class="jsx-785997980 "
      >
        <span>
          Rainbow Dash
        </span>
      </span>
    </span>
  </a>
</div>
`;

exports[`renders a starred, selected chip for a cached dashboard 1`] = `
<div>
  <a
    class="link"
    data-test="dashboard-chip"
    href="/rainbowdash"
  >
    <span
      class="jsx-3285323350 selected"
      data-test="dhis2-uicore-chip"
    >
      <span
        class="jsx-3489677657"
        data-test="dhis2-uicore-chip-icon"
      >
        <svg
          color="#ffffff"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.397 7.041l-6.051.88-.114.022a1 1 0 00-.44 1.683l4.378 4.268-1.033 6.027-.013.11a1 1 0 001.463.944L12 18.13l5.413 2.845.1.047a1 1 0 001.35-1.1l-1.034-6.028 4.38-4.268.078-.086a1 1 0 00-.633-1.62l-6.052-.879-2.705-5.484a1 1 0 00-1.794 0z"
            fill="currentColor"
          />
        </svg>
      </span>
      <span
        class="jsx-785997980 "
      >
        <span
          class="labelWithAdornment"
        >
          Rainbow Dash
        </span>
        <svg
          class="adornment selected"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m4.5 7v1h-1.5v6h10v-6h-1.5v-1h1.5c.5522847 0 1 .44771525 1 1v6c0 .5522847-.4477153 1-1 1h-10c-.55228475 0-1-.4477153-1-1v-6c0-.55228475.44771525-1 1-1zm5 3v1h-3v-1zm2.1130214-9.31661889.7739572.63323779-4.84985989 5.92760659-3.3906721-3.3906721.70710678-.70710678 2.60944661 2.60955339z"
            fill="inherit"
            fill-rule="evenodd"
          />
        </svg>
      </span>
    </span>
  </a>
</div>
`;

exports[`renders a starred, selected chip for non-cached dashboard 1`] = `
<div>
  <a
    class="link"
    data-test="dashboard-chip"
    href="/rainbowdash"
  >
    <span
      class="jsx-3285323350 selected"
      data-test="dhis2-uicore-chip"
    >
      <span
        class="jsx-3489677657"
        data-test="dhis2-uicore-chip-icon"
      >
        <svg
          color="#ffffff"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.397 7.041l-6.051.88-.114.022a1 1 0 00-.44 1.683l4.378 4.268-1.033 6.027-.013.11a1 1 0 001.463.944L12 18.13l5.413 2.845.1.047a1 1 0 001.35-1.1l-1.034-6.028 4.38-4.268.078-.086a1 1 0 00-.633-1.62l-6.052-.879-2.705-5.484a1 1 0 00-1.794 0z"
            fill="currentColor"
          />
        </svg>
      </span>
      <span
        class="jsx-785997980 "
      >
        <span>
          Rainbow Dash
        </span>
      </span>
    </span>
  </a>
</div>
`;

exports[`renders an unstarred chip for a non-offline dashboard 1`] = `
<div>
  <a
    class="link"
    data-test="dashboard-chip"
    href="/rainbowdash"
  >
    <span
      class="jsx-3285323350 "
      data-test="dhis2-uicore-chip"
    >
      <span
        class="jsx-785997980 "
      >
        <span>
          Rainbow Dash
        </span>
      </span>
    </span>
  </a>
</div>
`;

exports[`renders an unstarred chip for an offline dashboard 1`] = `
<div>
  <a
    class="link"
    data-test="dashboard-chip"
    href="/rainbowdash"
  >
    <span
      class="jsx-3285323350 "
      data-test="dhis2-uicore-chip"
    >
      <span
        class="jsx-785997980 "
      >
        <span
          class="labelWithAdornment"
        >
          Rainbow Dash
        </span>
        <svg
          class="adornment"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m4.5 7v1h-1.5v6h10v-6h-1.5v-1h1.5c.5522847 0 1 .44771525 1 1v6c0 .5522847-.4477153 1-1 1h-10c-.55228475 0-1-.4477153-1-1v-6c0-.55228475.44771525-1 1-1zm5 3v1h-3v-1zm2.1130214-9.31661889.7739572.63323779-4.84985989 5.92760659-3.3906721-3.3906721.70710678-.70710678 2.60944661 2.60955339z"
            fill="inherit"
            fill-rule="evenodd"
          />
        </svg>
      </span>
    </span>
  </a>
</div>
`;
